File: 3486.longest-special-path-ii.py
#
# @lc app=leetcode id=3486 lang=python3
#
# [3486] Longest Special Path II
#

# @lc code=start
from collections import defaultdict, Counter
class Solution:
    def longestSpecialPath(self, edges: List[List[int]], nums: List[int]) -> List[int]:
        n = len(nums)
        tree = defaultdict(list)
        for u, v, l in edges:
            tree[u].append((v, l))
            tree[v].append((u, l))

        self.max_len = 0
        self.min_nodes = float('inf')

        # Helper to check the special path constraint
        def is_valid(counter):
            duplicates = 0
            for count in counter.values():
                if count == 2:
                    duplicates += 1
                elif count > 2:
                    return False
            return duplicates <= 1

        def dfs(u, parent, path_counter, cur_len, cur_nodes):
            val = nums[u]
            path_counter[val] += 1

            if is_valid(path_counter):
                if cur_len > self.max_len:
                    self.max_len = cur_len
                    self.min_nodes = cur_nodes
                elif cur_len == self.max_len:
                    self.min_nodes = min(self.min_nodes, cur_nodes)
                for v, l in tree[u]:
                    if v == parent:
                        continue
                    dfs(v, u, path_counter, cur_len + l, cur_nodes + 1)
            # Backtrack
            path_counter[val] -= 1

        dfs(0, -1, Counter(), 0, 1)
        # After traversal, double-check that answer meets constraints
        # (already ensured by is_valid during traversal)
        return [self.max_len, self.min_nodes]
# @lc code=end

File: 3486.longest-special-path-ii.java
#
# @lc app=leetcode id=3486 lang=java
#
# [3486] Longest Special Path II
#
# @lc code=start
import java.util.*;
class Solution {
    public int[] longestSpecialPath(int[][] edges, int[] nums) {
        int n = nums.length;
        List<int[]>[] tree = new ArrayList[n];
        for (int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            tree[u].add(new int[]{v, w});
            tree[v].add(new int[]{u, w});
        }
        int[] result = new int[]{0, Integer.MAX_VALUE};
        // For each node, initiate a DFS to consider all downward paths from that node
        for (int i = 0; i < n; ++i) {
            dfs(i, -1, tree, nums, new HashMap<>(), false, 0, 1, result);
        }
        return new int[]{result[0], result[1]};
    }
    private void dfs(int node, int parent, List<int[]>[] tree, int[] nums, Map<Integer,Integer> count, boolean usedDuplicate, int pathLen, int nodeCount, int[] result) {
        count.put(nums[node], count.getOrDefault(nums[node], 0) + 1);
        boolean isDuplicate = count.get(nums[node]) == 2;
        if (count.get(nums[node]) > 2 || (usedDuplicate && isDuplicate)) {
            count.put(nums[node], count.get(nums[node]) - 1);
            return;
        }
        if (pathLen > result[0]) {
            result[0] = pathLen;
            result[1] = nodeCount;
        } else if (pathLen == result[0]) {
            result[1] = Math.min(result[1], nodeCount);
        }
        for (int[] nei : tree[node]) {
            int next = nei[0], weight = nei[1];
            if (next == parent) continue;
            dfs(next, node, tree, nums, count, usedDuplicate || isDuplicate, pathLen + weight, nodeCount + 1, result);
        }
        count.put(nums[node], count.get(nums[node]) - 1);
    }
}
# @lc code=end